src/unix/linux/disk.rs

```diff
@@ -83,39 +83,57 @@ impl DiskInner {
     }
 
-    fn efficient_refresh(&mut self, procfs_disk_stats: &HashMap<String, DiskStat>) -> bool {
-        let Some((read_bytes, written_bytes)) =
-            procfs_disk_stats.get(&self.actual_device_name).map(|stat| {
-                (
-                    stat.sectors_read * SECTOR_SIZE,
-                    stat.sectors_written * SECTOR_SIZE,
-                )
-            })
-        else {
-            sysinfo_debug!("Failed to update disk i/o stats");
-            return false;
-        };
+    fn efficient_refresh(
+        &mut self,
+        refresh_kind: DiskRefreshKind,
+        procfs_disk_stats: &HashMap<String, DiskStat>,
+    ) -> bool {
+        if refresh_kind.kind() && self.type_ == DiskKind::Unknown(-1) {
+            self.type_ = find_type_for_device_name(&self.device_name);
+        }
 
-        self.old_read_bytes = self.read_bytes;
-        self.old_written_bytes = self.written_bytes;
-        self.read_bytes = read_bytes;
-        self.written_bytes = written_bytes;
-
-        unsafe {
-            let mut stat: statvfs = mem::zeroed();
-            let mount_point_cpath = to_cpath(&self.mount_point);
-            if retry_eintr!(statvfs(mount_point_cpath.as_ptr() as *const _, &mut stat)) == 0 {
-                let tmp = cast!(stat.f_bsize).saturating_mul(cast!(stat.f_bavail));
-                self.available_space = cast!(tmp);
-                true
-            } else {
-                false
+        if refresh_kind.io_usage() {
+            if self.actual_device_name.is_none() {
+                self.actual_device_name = Some(get_actual_device_name(&self.device_name));
             }
+            let (read_bytes, written_bytes) = if let Some((read_bytes, written_bytes)) =
+                procfs_disk_stats
+                    .get(self.actual_device_name.as_ref().unwrap())
+                    .map(|stat| {
+                        (
+                            stat.sectors_read * SECTOR_SIZE,
+                            stat.sectors_written * SECTOR_SIZE,
+                        )
+                    }) {
+                (read_bytes, written_bytes)
+            } else {
+                sysinfo_debug!("Failed to update disk i/o stats");
+                (0, 0)
+            };
+
+            self.old_read_bytes = self.read_bytes;
+            self.old_written_bytes = self.written_bytes;
+            self.read_bytes = read_bytes;
+            self.written_bytes = written_bytes;
         }
+
+        if refresh_kind.details() {
+            let (total_space, available_space, is_read_only) =
+                match unsafe { load_statvfs_values(&self.mount_point, refresh_kind) } {
+                    Some((total, available, is_read_only)) => (total, available, is_read_only),
+                    None => (0, 0, false),
+                };
+
+            self.total_space = total_space;
+            self.available_space = available_space;
+            self.is_read_only = is_read_only;
+        }
```

Comment:
```suggestion
        if let Some((total_space, available_space, is_read_only)) =
            unsafe { load_statvfs_values(&self.mount_point, refresh_kind) }
        {
            self.total_space = total_space;
            self.available_space = available_space;
            self.is_read_only = is_read_only;
        }
```

Two things: `refresh_kind.details()` is already checked in `load_statvfs_values` and values should not be updated if the call failed (I think).

Reply:
Oh, you're right. IMO it would be more consistent with the rest of the code if `load_statvfs_values` didn't check `refresh_kind`, so I'll approach it slightly differently.

Reply:
I actually kinda like it. Like that you can't forget the check.

Reply:
I don't disagree, but I like the consistency of having all the `refresh_kind` checks in the higher-level refresh functions even more.

Reply:
As you prefer. Both ways work for me.